.pre-commit-config.yaml

```diff
@@ -3,22 +3,40 @@ repos:
     hooks:
       - id: web-js-lint-check
         name: Web JS Lint Check
-        entry: bash -c 'make -j$(nproc) -C webui/react check-eslint check-prettier-js'
+        entry: >
+          bash -c '
+            sh ./pre-commit/strip.sh "$0 $*" webui\/react\/ |
+            (
+              FILE_PATHS=$(cat);
+              make -j$(nproc) -C webui/react prettier PRE_ARGS="-- -c $FILE_PATHS" eslint ES_ARGS="$FILE_PATHS"
+            )
+          '
```

Comment:
I think we can improve this. :)

That's way too many hoops for getting the output of a script into a variable.  Just put the script in the command substitution instead.
```suggestion
          bash -c '
              FILE_PATHS=$(sh ./pre-commit/strip.sh "$0 $*" webui\/react\/);
              make -j$(nproc) -C webui/react prettier PRE_ARGS="-- -c $FILE_PATHS" eslint ES_ARGS="$FILE_PATHS"
          '
```

Same applies to the other two sections.  The subshell and pipe are all unnecessary.  Though, realistically, this is all a bit convoluted and challenging to read.  Why isn't this whole block just a stand-alone script in the ./pre-commit folder instead of being embedded in the yaml?  The whole difference in the code is just eslint v/s stylint v/s check-package-lock, and a corresponding variable name.  That could be one script with the first argument being the target and subsequent args being the files. :)  Heck, it could be one script with a couple of symlinks, where `$(basename $0)` in a case statement decides how it's called, and then all the arguments are files.  That might be excessive, though.

Incidentally, it might not be a terrible thing to depend on `realpath` for the path conversion; it's part of the GNU coreutils.  Check this out:
```shell
sauer@host:~/det $ realpath --relative-to webui/react/src webui/react/src/hooks/useTheme.ts 
hooks/useTheme.ts
```
It doesn't even depend on string matching; it uses the actual filesystem:
```shell
sauer@host:~/det $ realpath --relative-to webui/react/src ../determined/webui/./././react/src/hooks/useTheme.ts 
hooks/useTheme.ts
```

Reply:
For example, something like this:
```bash
#!/usr/bin/env bash
target=basename($0)
D=webui/react
files=$(realpath --relative-to="$D" "$@")
case $target in
  eslint)  make -j$(nproc) -C "$D" prettier PRE_ARGS="-- -c $files" $target ES_ARGS="$files"  ;;
  a)  make -j$(nproc) -C "$D" prettier PRE_ARGS="-- -c $files" $a A_ARGS="$files"  ;;
  b)  make -j$(nproc) -C "$D" prettier PRE_ARGS="-- -c $files" $b B_ARGS="$files"  ;;
  *) echo "unknonwn target '$target'"; exit 1 ;;
esac
```
Then you name is eslint and also make symlinks names a and b to support those three checks.  Or something like that.

Reply:
I didn't even know about `realpath`. That's exactly what I wanted. I'll check it out.

Reply:
It even takes multiple arguments, and returns them al converted l one per line. It's tailor made for this. :)

Reply:
Even though I actually didn't want to make shell script for pre-commit, I made it cuz i didn't know about `realpath` command.  Now, i think its more concise
What about doing this?
```yaml
 entry: >
          bash -c '
            FILE_PATHS=$(realpath --relative-to webui/react $0 $@) &&
            make -j$(nproc) -C webui/react prettier PRE_ARGS="-- -c $FILE_PATHS" eslint ES_ARGS="$FILE_PATHS"
          '
```

Reply:
Yeah I think that is probably OK. Do you need $0 in there, though?

I still would rather have it pulled out in a separate script even if it's just two lines. You get syntax hilighting in your editor and cleaner YAML that way. And no weird YAML quoting and indentation junk. :)

Reply:
> Yeah I think that is probably OK. Do you need $0 in there, though?

`$0` is necessary cuz Mac  uses bash v3. `$@` doesn't take `$0` in v3 (it does above v4), so we will miss one file without `$0`.

> I still would rather have it pulled out in a separate script even if it's just two lines. You get syntax hilighting in your editor and cleaner YAML that way. And no weird YAML quoting and indentation junk. :)

Can I make 3 separate files for js, css, and misc? I don't wanna write logic in a shell script cuz some are not familiar with shell script like me :) and it is painful to read shell script even for shell professional. Also, its more flexible to change code with separate files in the future

Reply:
$0  It's the name the script was called as, it's not one of the arguments. That's why I was asking if it needs to be included. The arguments start with $1.

You might echo $files and make sure this is doing what you think it's doing. :)

Reply:
Yes. I did `echo`. 
if its just `echo`, we can get all files. However, we don't get all args with `bash -c 'echo $@'` on pre-commit.

Reply:
I'll bet that'd behave more consistently with the script as  stand alone file as well. ;)

I suspect there's something going on with the way the exec gets called with the bash -c, variant, but I'd rather not speculate. If you pull it out to a separate script, $0 will be the script name and $@ should work in a less surprising way.

Reply:
Thanks for the advices. I made a shell script.